coverage/sqldata.py

```diff
@@ -615,6 +615,42 @@ def touch_files(self, filenames: Collection[str], plugin_name: Optional[str] = N
                     # Set the tracer for this file
                     self.add_file_tracers({filename: plugin_name})
 
+    def purge_files(self, filenames: Iterable[str], context: Optional[str] = None) -> None:
+        """Purge any existing coverage data for the given `filenames`.
+
+        If `context` is given, purge only data associated with that measurement context.
+        """
+
+        if self._debug.should("dataop"):
+            self._debug.write(f"Purging {filenames!r} for context {context}")
+        self._start_using()
+        with self._connect() as con:
+
+            if context is not None:
+                context_id = self._context_id(context)
+                if context_id is None:
+                    raise DataError("Unknown context {context}")
+            else:
+                context_id = None
+
+            if self._has_lines:
+                table = 'line_bits'
+            elif self._has_arcs:
+                table = 'arcs'
+            else:
+                return
+
+            for filename in filenames:
+                file_id = self._file_id(filename, add=False)
+                if file_id is None:
+                    continue
+                self._file_map.pop(filename, None)
```

Comment:
After I merged, I noticed this line: if you are only removing one context for the file, then we don't want to remove the file from the file_map, correct?  Why is it important to remove it from the map, and what should we do if it's only a single context being removed?

Reply:
It looks like the original `cur.execute` call didn't actually remove data because the cursor wasn't iterated. I replaced it with `cur.execute_void` and it had a better effect.

If you don't need the per-context operation, then let's definitely simplify it to only remove the entire file.  I think I'd have it remove the data for the file, but leave the file in the set of measured files, and state that it's possible for a "measured file" to have no data.  Would that work for you?

If we find a need for per-context purging later, we could add it.

Reply:
I've implemented this in https://github.com/nedbat/coveragepy/pull/1551 - wasn't sure how to add to this pull request so ended up with a new one.  Hope that is OK...